Add edit in staged payment

agnieszkajarosikloj opened the pull request:
## Description

This PR adds edit in staged payment. 
Edit in other payment types will be added in separate PRs.

**Acceptance Criteria** 

- the ability to change the recipient
- the ability to change the value of the staged payment and the token selected
- the ability to change the name of a milestone
- the ability to change the percent allocation of a milestone
- the ability to add and remove milestones

<img width="488" alt="Screenshot 2022-10-05 at 12 30 16" src="https://user-images.githubusercontent.com/91876137/194040132-ee41d51f-cd3b-463f-9740-ca2d803b9354.png">


Resolves  #3795

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/JoinColony/colonyDapp?pullRequest=3947) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/JoinColony/colonyDapp?pullRequest=3947) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/JoinColony/colonyDapp?pullRequest=3947) it.</sub>

willm30 reviewed (changes requested):
Nice stuff. Got a few notes. 

I think the heading for milestone could read: "Change milestone: 1" because all the other headings have "Change" in them. 

Also, do we need it to read "Change to", instead of just "To"? We could avoid repetition of "change" since it's included in the heading. (Would apply to them all, in all edit modals)

![editstaged](https://user-images.githubusercontent.com/64402732/195289914-ccd8ecfd-1a61-4528-9af1-0c16e70a956b.png)

I'm noticing this behaviour when scrolling within the milestone input. This can be fixed if you increase the height of the input slightly. 
![milestone](https://user-images.githubusercontent.com/64402732/195290863-b3408009-d422-405d-a177-8959d4f6fe3c.gif)

The space between the token symbol and the text is 10px in figma. 
![spac](https://user-images.githubusercontent.com/64402732/195292399-815d2966-6ec5-4fdd-bd94-07d2035bca01.png)

The selected recipient for To is missing the "@" before the name: 
![missingat](https://user-images.githubusercontent.com/64402732/195292970-a6ffe935-634a-4a10-8755-220550a5f4f3.png)

And the cursor when hovering over the edit icon should be `cursor: pointer` whereas it's currently default.

agnieszkajarosikloj commented:
@willm30 I added fixes:

- changed headers and "To" label, 
- added fixes to styles,
- changed label in UserPickerWithSearch component -  added "@" before the name.

Can you re-review please?

chinins reviewed (changes requested):
Mostly all good, but there is one part that needs refactoring in my opinion. Please, let me know once done and I'll re-review straight away.

agnieszkajarosikloj commented:
@chinins I have added changes. Can you have a look please?

chinins reviewed (approved):
Thank you - that's so much better! Just, please, wrap the 2 variables in useMemo (as per my direct comments).

jakubcolony commented on src/modules/pages/components/ExpenditurePage/utils.ts:
```diff
@@ -312,6 +312,66 @@ export const updateValues = (values, confirmedValues) => {
+        return true;
+      }
+      return confirmedValues.staged.milestones?.every(
+        (confirmedMielstone) => confirmedMielstone.id !== milestone.id,
```
Typo

jakubcolony replied:
Looks like this one slipped through your fix @agnieszkajarosikloj

jakubcolony commented on src/modules/pages/components/ExpenditurePage/utils.ts:
```diff
@@ -312,6 +312,66 @@ export const updateValues = (values, confirmedValues) => {
+      ...sameMilestones,
+      ...changedMilestones?.map((milestone) => {
+        const newValue = confirmedValues.staged.milestones?.find(
+          (confirmedMielstone) => confirmedMielstone.id === milestone.id,
```
Typo

jakubcolony reviewed (changes requested):
Just spotted a typo in the screenshot Will posted earlier so I highlighted all of its occurrences

jakubcolony commented on src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedStaged/ChangedMilestone/ChangedMilestone.tsx:
```diff
@@ -0,0 +1,72 @@
+
+import ChangeItem from '../../ChangedMultiple/ChangeItem';
+import { ValueOf } from '../../ChangedValues/ChangedValues';
+
```
Can remove the empty line?

jakubcolony reviewed (approved):
Nice work @agnieszkajarosikloj! 

Just noticed there's one typo left

agnieszkajarosikloj commented:
@willm30 This is ready for re-review.

willm30 reviewed (commented):
Where it says "Change Staged", should that read "Change Token"? That was the change I made on that line. 

![editp](https://user-images.githubusercontent.com/64402732/196138798-560cdcea-2708-4e3a-b7f8-2fcd9884e6eb.png)

agnieszkajarosikloj commented:
@willm30 I have changed header. I left one header for amount and user. Can you have a look?


<img width="479" alt="Screenshot 2022-10-17 at 11 56 23" src="https://user-images.githubusercontent.com/91876137/196148437-a0f2e682-f190-4442-9042-edba760834ed.png">

willm30 commented:
@agnieszkajarosikloj Would it be possible to put the "Changed User" on its own line? Also, the user avatar looks to be larger than the token icons. I think we could make these all the same size.

agnieszkajarosikloj commented:
@willm30 headers and styles have been changed. Can you have a look please?

willm30 reviewed (approved):
Nice work! One final request (sorry for the back and forth): I think we can omit "Staged" from "Change Staged Amount" and "Change Staged User" to leave "Change Amount" and "Change User". Thereby maintaining consistency between headers. 

And also, the "title" and "description" in "Change title" and "Change description" aren't capitalised like in other headers. Could we update that too?
